### PR TITLE
[Triton] [Inductor] Enable TMA store for TMA mm templates

### PR DESCRIPTION
**Pull request by njriasan**
Summary:
Adds support for TMA store in all TMA matmul templates (notably persistent_tma including addmm and scaled_mm). This works by requiring a template be registered with `tma_store=True` and when met constructs indices/range_trees to hook into the existing code base's TMA store support.

This also includes a couple notable changes:
- Adds support in the TMA template support for checking the output layout.
- Adds support for "hoisting" the tensor descriptor to the top of the kernel. This will currently only be used by template code right now, but in principle it can be generalized to other implementation.
- Supports considering multiple indices as the "contiguous" index. This is handled with support for transposing the input data when the alignment is no longer consistent. In general since the TMA support is derived from the index it doesn't seems reasonable that the 1D index math forces a certain alignment depending on index ordering so long as the layout matches.

Test Plan:
Tested with test_max_autotune.py unit tests.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @mlazos

### CODE REVIEW

File: torch/_inductor/select_algorithm.py [hunk 1]
```diff
@@ -1018,26 +1093,74 @@ def store_output(
...
+                output_layout = self.output_node.get_layout()
+                needs_transpose = not self.output_node.get_layout().is_contiguous()
+                if needs_transpose:
+                    output_index = index_symbols[::-1]
+                else:
+                    output_index = index_symbols
+
+                assert len(tma_info.block_sizes) == len(output_index), "Unexpected number of indices"
+                assert len(tma_info.shapes) == len(output_index), "Unexpected number of indices"
+                # TODO: Fix all of the naming logic.
```
Comment (njriasan) [inline torch/_inductor/select_algorithm.py]: I need to cleanup this code a lot, but the biggest component here is that I'm generating the remapping from TMA indices to regular pointer indices (happy to split this off into a helper). To do this I'm generating a lot of intermediate names and I'm not familiar with the ideal way of doing that in this code base (a lot of the helpers seems to have side effects).

I'll keep looking around/ask an llm, but any pointers are appreciates (cc: @eellison). This code is now fully working for fusions in addmm.

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -1665,11 +1667,31 @@ def can_use_tma(*matrices: IRNode, add_guards: bool = False) -> bool:
...
-        sizes = x.get_size()
-        strides = x.get_stride()
+    def _is_tma_compatible_layout(layout: Optional[Layout]) -> bool:
+        if layout is None:
+            return True
+        sizes = layout.size
+        strides = layout.stride
+        dtype = layout.dtype
+
+        # TODO: Check Base pointer 16-byte aligned for output?
```
Comment (njriasan) [inline torch/_inductor/utils.py]: Please let me know if you have thoughts on how I can check this. Also happy to do this as a followup.

**Review by eellison — COMMENTED** [review]
Could we add this logic to [triton codegen store](https://github.com/pytorch/pytorch/blob/179511694cbbc22be87e5fa4b6965e41129032d1/torch/_inductor/codegen/triton.py#L2609) instead ? 

Potentially with a flag on the kernel to prefer tma stores.

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -2121,7 +2173,6 @@ def indexing(
             and not self._load_mask
             and len(mask_vars - dense_mask_vars) == 0
             and not self.is_indirect_indexing(index)
-            and have_loop_vars
```
Comment (njriasan) [inline torch/_inductor/codegen/triton.py]: Its not clear to me why this was previously a requirement. It doesn't apply in my indexing example and I believe my example is a clear case where TMA should be supported. Please let me know if this needs a further look.

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -2304,31 +2355,44 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
...
                     )
                     if not tma_compatibility_checker.are_block_parameters_compatible(
-                        options.params
+                        block_params
                     ):
-                        return None
+                        # Some inputs may intend to store the values as (X, Y),
+                        # not (Y, X). As a result, try and tranpsose the blocks
+                        #
+                        block_params = block_params.transpose()
```
Comment (njriasan) [inline torch/_inductor/codegen/triton.py]: This is an important generalization of TMA support. We should be able to store the result if it is contiguous as X, Y or Y, X.

File: torch/_inductor/codegen/triton.py [hunk 5]
```diff
@@ -2451,23 +2517,45 @@ def codegen_block_ptr(
...
+        #       happens we will generate a tl.trans().
+        #    3. A mismatched provided shape. When this occurs we will error.
+        #    4. No shape is provided. This will proceed with the default explicit broadcast
+        #       described below.
+        #
+        # To prevent unintended side effects we will gate options 1-3 behind isinstance(indexing, TensorDescriptorOptions).
+        if isinstance(indexing, TensorDescriptorOptions) and value.shape:
+            str_final_shape = [symt.name for symt in indexing.final_shape]
+            if value.shape[::-1] == str_final_shape:
+                value = f"tl.trans({value})"
```
Comment (njriasan) [inline torch/_inductor/codegen/triton.py:2545]: I don't know how to generate a test case for this constraint because its not clear how I can set the output layout. By changing the default output layout I hit a bug elsewhere in the code and then afterwards was able to verify it worked manually.

File: torch/_inductor/codegen/triton.py [hunk 6]
```diff
@@ -3759,6 +3851,18 @@ def cse_multiple(line, broadcasted_values, masks, dtypes):
 
         return tuple(result_vars)
 
+    def codegen_prologue(self, code: IndentedBuffer):
+        """
+        Generate the output from prologue. This should be
```
Comment (njriasan) [inline torch/_inductor/codegen/triton.py:3871]: This is needed to enable extracting the make_tma_descriptor. Right now I assume if you generate from a kernel and force TMA this is safe, otherwise its not.

File: torch/_inductor/select_algorithm.py [hunk 7]
```diff
@@ -1014,6 +1030,66 @@ def hook():
 
         return self._register_hook(hook_key, hook)
 
+    def _generate_index_from_tma_index(
```
Comment (njriasan) [inline torch/_inductor/select_algorithm.py:1041]: This logic formats the index in a compatible way from any template to correspond to other inductor assumptions. All of the tma specific code lives in indexing, but reconstructing a valid index is necessary.

File: torch/_inductor/select_algorithm.py [hunk 8]
```diff
@@ -1045,26 +1122,72 @@ def store_output(
...
-            )
-            self.template_mask = mask
+            output_layout = self.output_node.get_layout()
             self.template_out = val
-            self.template_indices = indices
-            output_index = self.output_node.get_layout().make_indexer()(index_symbols)
-            output_index = self.rename_indexing(output_index)
-            if output_index == contiguous_index:
-                output_index = sympy.Symbol("xindex", integer=True)
+            if self.tma_store:
```
Comment (njriasan) [inline torch/_inductor/select_algorithm.py]: Some additional work is likely needed to support making tma_store autotunable, but it should be achievable if you just map everything to `xindex` as before.

File: torch/_inductor/codegen/triton.py [hunk 9]
```diff
@@ -2304,31 +2355,44 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
...
-                        return None
+                        # Some inputs may intend to store the values as (X, Y),
+                        # not (Y, X). As a result, try and tranpsose the blocks
+                        #
+                        block_params = block_params.transpose()
+                        if (
+                            not tma_compatibility_checker.are_block_parameters_compatible(
+                                block_params
+                            )
+                            and copy_shape is not None
```
Comment (njriasan) [inline torch/_inductor/codegen/triton.py]: This may not be a sufficient check that we know val's shape.

File: torch/_inductor/kernel/mm.py [hunk 10]
```diff
@@ -776,13 +783,14 @@ def tuned_mm(mat1, mat2, *, layout=None):
...
 
-        if use_triton_tma_template(mat1, mat2):
+        if use_triton_tma_template(mat1, mat2, output_layout=layout):
             # Get TMA template params using the new unified function
             for kwargs, extra_kwargs in V.choices.get_mm_configs(
                 kernel_inputs, layout, persistent_tma_mm_template.name, "mm"
             ):
                 persistent_tma_mm_template.maybe_append_choice(
                     choices,
+                    tma_store = not kwargs["TMA_EXPERIMENTAL_API"],
```
Comment (NikhilAPatel) [inline torch/_inductor/kernel/mm.py]: If I'm understanding this correctly, this will essentially turn on TMA store by default (as long as we're running with a version of Triton with the new API, `val_shape` is supplied in the kernel, and the inputs satisfy the TMA constraints). Based on the various investigations done on the performance of TMA store, I think this is too generous. We may want to just keep this `False` by default until we add in the ability to autotune over this.

**Review by NikhilAPatel — APPROVED** [review]
Just had one comment above about the decision to enable TMA store by default, but otherwise looks good to me! Very impressive work thanks for tackling this!

File: torch/_inductor/utils.py [hunk 11]
```diff
@@ -1679,23 +1681,44 @@ def can_use_tma(*matrices: IRNode, add_guards: bool = False) -> bool:
...
             return False
 
         # dtype ∈ {FP16, BF16, FP8-E4M3FN}
-        if dtype not in (torch.float16, torch.bfloat16, torch.float8_e4m3fn):
-            return False
-
-        # Base pointer 16-byte aligned
-        if x.get_name() in V.graph.unaligned_buffers:
+        if dtype not in (torch.float16, torch.bfloat16, torch.float8_e4m3fn) and (
+            not allow_float32 or dtype != torch.float32
```
Comment (njriasan) [inline torch/_inductor/utils.py:1729]: allow_float32 shouldn't be necessary in general, but this PR has already gotten quite large, so I want to reduce the chance of a revert. This is necessary because some of the fp8 examples have float32 outputs. In general I see no reason float32 shouldn't be supported.

File: torch/_inductor/codegen/triton.py [hunk 12]
```diff
@@ -2304,31 +2356,49 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
...
                     tma_compatibility_checker = cast(
                         TMACompatibilityChecker, tma_compatibility_checker
                     )
                     if not tma_compatibility_checker.are_block_parameters_compatible(
-                        options.params
+                        block_params
                     ):
-                        return None
+                        # Some inputs may intend to store the values as (X, Y),
+                        # not (Y, X). As a result, try and tranpsose the blocks
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py]: Do you have any examples of this ? And when they would be tma compatible as Y, X but not X, Y ?

File: torch/_inductor/codegen/triton.py [hunk 13]
```diff
@@ -2451,23 +2523,45 @@ def codegen_block_ptr(
...
-        # These dims no longer need broadcasting.
-        for idx, (dim, broadcast_dim) in enumerate(
-            zip(indexing.final_shape, indexing.broadcast_shape)
-        ):
-            if V.graph.sizevars.statically_known_equals(dim, broadcast_dim):
-                indexing.broadcasting_dims[idx] = False
+        # TMA stores may require transposing the data to ensure we are contiguous along
+        # the final dimension. We do this by checking the shape information on value.
+        # It can either
+        #    1. Match the final shape. In this case no broadcast/reshape
```
Comment (eellison) [inline torch/_inductor/codegen/triton.py:2532]: Is this different from other triton stores ? In triton, you can store [X, Y] and it can be coalesced on either dimension and you do not need to manually transpose.

File: torch/_inductor/select_algorithm.py (hunk 8 above)
Comment (eellison) [reply torch/_inductor/select_algorithm.py]: Couldnt this be a flag on the template?

File: torch/_inductor/utils.py (hunk 2 above)
Comment (eellison) [reply torch/_inductor/utils.py]: Check if `layout.offset` is statically divisible by 16

**Review by eellison — COMMENTED** [review]
Looks good, a few questions. Also cc @blaine-rister for block pointer review

File: torch/_inductor/codegen/triton.py (hunk 12 above)
Comment (njriasan) [reply torch/_inductor/codegen/triton.py]: So in general the TMA storage always has to be contiguous in the last dimension, which is the requirement this seeks to satisfy. By default imagine we seek to store a block with shapes `[BLOCK_M, BLOCK_N]`. If the output layout (or reused storage) is contiguous in N, then this is fine and works as intended. If however it were contiguous in M, not M, then the TMA layout would require a transposed allocation, something like:
```
desc = make_tensor_descriptor(
    shape=[N, M],
    ....
)
...
desc.store((n_off, m_off), tl.trans(data))
```

This is the extreme case that should supported for generality.

However, inductor is also just determining if the shape is `[N, M]` (and not [M, N]) based on the index construction. In particular, the index decomposition logic iterates through the indices in sorted order, meaning it does Y then X, resulting in always checking if X is inner dimension for the purposes of TMA support. This doesn't match the intention of the actual template store, which is the [M, N] shape.

I should also note that I believe this is a general gap in the existing TMA support, but without being certain you know the output shape of val its difficult to robustly support this case, which is the reason for the val shape restriction.

File: torch/_inductor/codegen/triton.py (hunk 13 above)
Comment (njriasan) [reply torch/_inductor/codegen/triton.py:2532]: Yes. See my comment above that for TMA in particular your shape always has the be contiguous on the innermost dimension and this transpose is done to satisfy that constraint.

File: torch/_inductor/kernel/mm.py (hunk 10 above)
Comment (njriasan) [reply torch/_inductor/kernel/mm.py]: Extremely reasonable feedback. I'll gate this behind an environment variable and we can enable it once we have better insights into TMA store's performance. BTW I did some investigation with a different kernel related to TMA load and I suspect that someone needs to do a deeper dive to understand the difference between the impact of `TMA store` the hardware feature and `TMA store` the Triton implementation.

File: torch/_inductor/select_algorithm.py (hunk 8 above)
Comment (njriasan) [reply torch/_inductor/select_algorithm.py]: Yes absolutely and given the feedback to disable TMA store by default I will need to do so. The "support" that was needed is actually if someone passes in values that indicate a TMA store index but they don't actually use TMA. This is purely an infra issue and not conceptual.

**Comment by njriasan** [comment]
@eellison I applied your feedback and @NikhilAPatel's feedback to disable the TMA_STORE by default. @NikhilAPatel now the same code works for TMA and no TMA, so you can now autotune over the TMA_STORE.

File: torch/_inductor/select_algorithm.py [hunk 14]
```diff
@@ -1045,26 +1130,134 @@ def store_output(
...
+                    # Reverse the index symbols because TMA is indexed
+                    # as (x, y) whereas the variables will naturally be indexed
+                    # as (y, x)
+                    epilogue_index_symbols.reverse()
+                else:
+                    mask_vars: list[str] = []
+                    for i, (index, val_shape) in enumerate(
+                        zip(index_symbols, val_shape)
+                    ):
+                        index_name = f"_tmpindex{i}"
```
Comment (njriasan) [inline torch/_inductor/select_algorithm.py]: Forgot to make these generic. TODO: Make these generic.

File: torch/_inductor/select_algorithm.py [hunk 15]
```diff
@@ -1014,13 +1036,78 @@ def hook():
...
+            num_dims (int): The total number of dimensions in the output.
+            block_name (Optional[str]): The name of the block variable. If not passed
+                in then we aren't reusing standard symbol names.
+
+        Returns:
+            list[str]: The lines used to generate the index.
+
+        """
+        if block_name:
+            # Generate the expected names for the structure:
```
Comment (blaine-rister) [inline torch/_inductor/select_algorithm.py:1070]: cc @jazlyn5: this could also be used for block ptr

File: torch/_inductor/codegen/triton.py [hunk 16]
```diff
@@ -2688,6 +2784,10 @@ def store(
             tma_compatibility_checker = self.tma_compatibility_checker_cls(
                 self, dtype, for_store=True
             )
+        if mode == "tma":
```
Comment (blaine-rister) [inline torch/_inductor/codegen/triton.py]: nit: Do we need to add another field / class for this? The functionality seems very similar to the existing tma checker, so I'm wondering if there's a way to consolidate the code. We might consider passing an option like `force=True` to the existing tma checker, or it might make sense to use an inner class like the following.
```suggestion
        if mode == "tma":
             class ForceTmaChecker(self.tma_compatibility_checker_cls):
                 # Inner class to override certain methods
                 def can_use_tma(self):
                     return True
              tma_compatibility_checker = ForceTmaChecker(
```

cc @kundaMwiza

File: torch/_inductor/codegen/triton.py [hunk 17]
```diff
@@ -336,7 +340,12 @@ def remove_dims(it):
 
         # Drop removable dimensions from the input.
         params = BlockParameters(
-            **{key: remove_dims(val) for key, val in dataclasses.asdict(params).items()}
+            **{
+                key: remove_dims(val)
+                for key, val in dataclasses.asdict(params).items()
+                if key != "is_transposed"
```
Comment (blaine-rister) [inline torch/_inductor/codegen/triton.py]: Why do we need this special case? Does transposition break something related to dim removal?

File: torch/_inductor/codegen/triton.py [hunk 18]
```diff
@@ -1641,6 +1658,15 @@ def __add__(self, other: BlockParameters) -> BlockParameters:
         a, b = tuple(dataclasses.asdict(x) for x in (self, other))
         return cls(**{key: a[key] + b[key] for key in a})
 
+    def transpose(self) -> BlockParameters:
```
Comment (blaine-rister) [inline torch/_inductor/codegen/triton.py:1663]: Is this telling us to read/write in one memory layout, then transpose it immediately afterwards? I like the idea, but I'm wondering if we could generalize the implementation to handle arbitrary permutations. Something like this might work.
```suggestion
    def transpose(self) -> BlockParameters:
           return BlockParameters(
               key: permute(val, self.permutation) # self.permutation is a sequence of integers that could be passed as an argument to torch.permute
           ) for key, val in dataclasses.asdict(self).items()
```

A good test case would be 3D loads/stores with permutations. There's an example here. https://github.com/pytorch/pytorch/blob/main/test/inductor/test_torchinductor_strided_blocks.py#L1041

Or is this too much for one PR? It could be a useful follow-up.

File: torch/_inductor/codegen/triton.py [hunk 19]
```diff
@@ -2304,31 +2358,49 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
...
                     tma_compatibility_checker = cast(
                         TMACompatibilityChecker, tma_compatibility_checker
                     )
                     if not tma_compatibility_checker.are_block_parameters_compatible(
-                        options.params
+                        block_params
                     ):
-                        return None
+                        # Some inputs may intend to store the values as (X, Y),
+                        # not (Y, X). As a result, try and tranpsose the blocks
```
Comment (blaine-rister) [inline torch/_inductor/codegen/triton.py]: It seems like this takes a brute force force approach to checking for permutations: try to transpose, and if it works, go ahead. I'm wondering if it would be possible to get the permutation map from the stride order instead. That approach could generalize to N dimensions in O(N) time, and might be a bit easier to follow. https://github.com/pytorch/pytorch/blob/92f9ed7ac37be57ec883e1cca798654b5b581607/torch/_inductor/ir.py#L341

N.B.: this method has some limitations around dynamic shapes. Namely, it requires passing the `shape_env` and is restricted to backed symints.

cc @nandesuka and @shunting314 used a similar method to pad permuted tensors. https://github.com/pytorch/pytorch/blob/92f9ed7ac37be57ec883e1cca798654b5b581607/torch/_inductor/ir.py#L3756-L3760

File: test/inductor/test_max_autotune.py [hunk 20]
```diff
@@ -314,11 +375,13 @@ def mm(a, b):
     @parametrize("a_transposed", (False, True))
     @parametrize("b_transposed", (False, True))
     @parametrize("dynamic", (False, True))
+    @parametrize("tma_store", (False, True))
     def test_max_autotune_addmm_persistent_tma(
```
Comment (blaine-rister) [inline test/inductor/test_max_autotune.py:382]: Do we have any 3D test cases? As noted elsewhere, it might also be good to implement and test support for arbitrary ND permutations.

**Review by blaine-rister — COMMENTED** [review]
Very cool and useful PR. The existing logic seems reasonable, but I left a few nits about opportunities to consolidate the code, or edge cases we might want to explore. Also, it might be possible to generalize this to ND without much added complexity. See the detailed comments above.

cc @kundaMwiza  who implemented automatic TMA codegen.

File: torch/_inductor/codegen/triton.py (hunk 19 above)
Comment (blaine-rister) [reply torch/_inductor/codegen/triton.py]: Another edge case to test would be [1,1] tensors: are they transposed or left alone? It could be good to add a CI test for that, although I'm not unsure whether TMA restrictions would get in the way of actually running it on GPU.

File: torch/_inductor/codegen/triton.py (hunk 16 above)
Comment (njriasan) [reply torch/_inductor/codegen/triton.py]: I think `force=True` is reasonable. I wanted to avoid situations where people basically agreed there was some "baseline" check that was always needed, but I actually prefer the update to the existing class. Thanks for the feedback.

File: test/inductor/test_max_autotune.py (hunk 20 above)
Comment (njriasan) [reply test/inductor/test_max_autotune.py:382]: No we don't because the GEMM Triton Templates don't support 3D by default. I think that would be a really great followup PR, perhaps for the bmm template. cc @NikhilAPatel

File: torch/_inductor/codegen/triton.py (hunk 17 above)
Comment (njriasan) [reply torch/_inductor/codegen/triton.py]: `remove_dims` assumes that `val` is an iterable. Since `is_transposed` is a boolean it will break dim removal.

File: torch/_inductor/codegen/triton.py (hunk 18 above)
Comment (njriasan) [reply torch/_inductor/codegen/triton.py:1663]: I agreed with this idea, but I think it should be done in a followup PR. Since this is primarily used for TMA store at this stage some followup thoughts will be needed for how to reliably test this for all the possible combinations.

Also this PR is already quite large and so selfishly I'd like to limit the scope.

File: torch/_inductor/codegen/triton.py (hunk 19 above)
Comment (njriasan) [reply torch/_inductor/codegen/triton.py]: The [1, 1] tensor is blocked by the TMA restrictions. In particular since all strides but 1 aren't divisible by 16 it can never be selected.

I agree this can definitely be generalized. However, I think we should do that in a followup PR if possible.

**Review by njriasan — COMMENTED** [review]
@blaine-rister Thanks for the feedback! I think we should do the ND generalization in a followup PR, but overall I agree with and appreciate the feedback.

File: torch/_inductor/codegen/triton.py (hunk 17 above)
Comment (blaine-rister) [reply torch/_inductor/codegen/triton.py]: Makes sense! So this still supports removing dims in a transposed block ptr, just not removing the boolean flag.

File: torch/_inductor/codegen/triton.py [hunk 21]
```diff
@@ -2304,31 +2358,49 @@ def match_block_expr() -> Optional[BlockDescriptorOptions]:
...
 
+                can_lift = False
                 if options_class == TensorDescriptorOptions:
                     nonlocal tma_compatibility_checker
                     tma_compatibility_checker = cast(
                         TMACompatibilityChecker, tma_compatibility_checker
                     )
                     if not tma_compatibility_checker.are_block_parameters_compatible(
-                        options.params
+                        block_params
```
Comment (kundaMwiza) [inline torch/_inductor/codegen/triton.py]: Hm, what is the reasoning for checking the original params rather than the simplified params (options.create removes broadcast and singleton dims). This might break a few tests, see https://github.com/pytorch/pytorch/pull/161602

File: torch/_inductor/codegen/triton.py (hunk 4 above)
Comment (kundaMwiza) [reply torch/_inductor/codegen/triton.py]: Related to the above comment. Perhaps we can perform this transpose in an override, `TensorDescriptorOptions.create`?

File: torch/_inductor/codegen/triton.py (hunk 21 above)
Comment (njriasan) [reply torch/_inductor/codegen/triton.py]: This was because I wanted to delay the options creation until I knew the inner most dimension. Based on your comment below I agree this is more suited to be implemented in `TensorDescriptorOptions.create`, which should avoid breaking any tests. I'll update this PR.

Thanks for the feedback!

Comment (kundaMwiza) [reply torch/_inductor/codegen/triton.py]: Oh okay. I did have a draft of sorting `block_params.strides` in descending order and storing a `permute_dims` attr on `TensorDescriptorOptions`. The ptr can then be loaded in a contiguous manner and the value transposed back to the kernel compute shape after. This sounds like what @blaine-rister is [referring](https://github.com/pytorch/pytorch/pull/160480/files#r2342600052) to but I thought it might hurt performance. But yeah, we can do that after this PR has landed

File: torch/_inductor/kernel/mm.py [hunk 22]
```diff
@@ -1149,8 +1174,17 @@ def tuned_scaled_mm(
         overriders = dict(USE_FAST_ACCUM=use_fast_accum)
 
         # TODO (paulzhan): There is no template that exists for bias and TMA
-        # Don't run tma template currently if bias exists
-        if use_triton_tma_template(mat_a, mat_b) and not bias:
+        # Don't run tma template currently if bias exist
+        if (
+            use_triton_tma_template(
```
Comment (coconutruben) [inline torch/_inductor/kernel/mm.py]: nit: can you move the layout thing inside the `use_triton_tma_template`

File: torch/_inductor/template_heuristics/triton.py [hunk 23]
```diff
@@ -1887,6 +1887,26 @@ def __init__(self) -> None:
...
+        kernel_inputs: KernelInputs,
+        op_name: str,
+    ) -> Generator[dict[str, Any], None, None]:
+        """
+        Generate scaled TMA template configs with both scaled MM and TMA-specific options.
+        """
+        # Get base scaled MM template configs from superclass
+        for template_kwargs in super()._get_template_configs_impl(
+            kernel_inputs,
+            op_name,
```
Comment (coconutruben) [inline torch/_inductor/template_heuristics/triton.py]: can this be a part of the existing TMA mixins, rather than the finalized mixins?

Comment (njriasan) [reply torch/_inductor/template_heuristics/triton.py]: Done.
